src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs

```diff
@@ -46,44 +46,40 @@ public LaunchBrowserFilter()
 
         public async ValueTask ProcessAsync(DotNetWatchContext context, CancellationToken cancellationToken)
         {
-            if (_suppressLaunchBrowser)
-            {
-                return;
-            }
-
             if (context.Iteration == 0)
             {
                 _reporter = context.Reporter;
 
-                if (CanLaunchBrowser(context, out var launchPath))
+                if (!_suppressLaunchBrowser && CanLaunchBrowser(context, out var launchPath))
                 {
                     context.Reporter.Verbose("dotnet-watch is configured to launch a browser on ASP.NET Core application startup.");
-                    _canLaunchBrowser = true;
+                    _shouldRefreshOnChanges = true;
                     _launchPath = launchPath;
                     _cancellationToken = cancellationToken;
 
                     // We've redirected the output, but want to ensure that it continues to appear in the user's console.
                     context.ProcessSpec.OnOutput += (_, eventArgs) => Console.WriteLine(eventArgs.Data);
                     context.ProcessSpec.OnOutput += OnOutput;
+                }
 
-                    if (!_suppressBrowserRefresh)
-                    {
-                        _refreshServer = new BrowserRefreshServer(context.Reporter);
-                        var serverUrl = await _refreshServer.StartAsync(cancellationToken);
+                if (!_suppressBrowserRefresh)
+                {
+                    _shouldRefreshOnChanges = true;
+                    _refreshServer = new BrowserRefreshServer(context.Reporter);
+                    var serverUrl = await _refreshServer.StartAsync(cancellationToken);
 
-                        context.BrowserRefreshServer = _refreshServer;
+                    context.BrowserRefreshServer = _refreshServer;
 
-                        context.Reporter.Verbose($"Refresh server running at {serverUrl}.");
-                        context.ProcessSpec.EnvironmentVariables["ASPNETCORE_AUTO_RELOAD_WS_ENDPOINT"] = serverUrl;
+                    context.Reporter.Verbose($"Refresh server running at {serverUrl}.");
+                    context.ProcessSpec.EnvironmentVariables["ASPNETCORE_AUTO_RELOAD_WS_ENDPOINT"] = serverUrl;
 
-                        var pathToMiddleware = Path.Combine(AppContext.BaseDirectory, "middleware", "Microsoft.AspNetCore.Watch.BrowserRefresh.dll");
-                        context.ProcessSpec.EnvironmentVariables["DOTNET_STARTUP_HOOKS"] = pathToMiddleware;
-                        context.ProcessSpec.EnvironmentVariables["ASPNETCORE_HOSTINGSTARTUPASSEMBLIES"] = "Microsoft.AspNetCore.Watch.BrowserRefresh";
-                    }
+                    var pathToMiddleware = Path.Combine(AppContext.BaseDirectory, "middleware", "Microsoft.AspNetCore.Watch.BrowserRefresh.dll");
```

Comment:
Thanks @zbecknell for the PR. The worrying part of this change is that it will by default inject an ASP.NET Core middleware in every app that's launched by `dotnet watch`. If we were to consider a change, we need a better signal to indicate that you want the refresh but not the browser launching. The env variable isn't well advertised so I wouldn't consider that as a reasonable option.